### ISSUE TITLE
Unable to create virtual joystick without actual joystick plugged in

### ISSUE THREAD

**Issue by kcgen** [dosbox-staging/dosbox-staging#2474]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

7dbf5cece42e9d2646522e2d1d3f57a2aeefabb2

### What Operating System are you using?

Linux x86_64

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

For games that require joystick input, the virtual joystick can be mapped to fixed keyboard events:

![2023-05-12_19-23](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/b3ffc24d-8a3b-45ce-873d-56a16fe5b0ad)

This means that even without a physical joystick plugged in, the game can still be played with consistent and repeatable keystrokes. (unlike with a physical controller, which often have inconsistent axis, hat, and button mappings).

This feature /does/ exists out of the box in DOSBox SVN and did in Staging prior to https://github.com/dosbox-staging/dosbox-staging/pull/154.  Unfortunately that PR unknowingly eliminated this ability.

#### Prerequisite:

 - Unplug all joysticks and controllers.

#### Explain how to reproduce:

1. Extract and launch [bcquest.zip](https://github.com/dosbox-staging/dosbox-staging/files/11468755/bcquest.zip)
5. The game's menu will appear.
2. Open the mapper: Ctrl (or CMD) + F1
3. Using your mouse, click the mapper's left-joystick's `Y+`, click the 'add bind' button, then using your keyboard, press the "up" arrow key to map it. Repeat this for `Y-` (and the down-arrow key), and finally the `X+` and `X-` (mapped to left and right arrow keys).
4. Click 'Save', then 'Exit' to close the matter and return to the game.
6. Press `1` on your keyboard to start a 1-player game. 
7. When the game starts, notice that your player repeatedly jumps and does not honor any keystrokes (uparrow and downarrow).

#### How it should work:

1. `git checkout 7dbf5cec^`
2. `git clean -fdx && ./autogen.sh && CC="ccache gcc" CXX="ccache g++" ./configure && make -j12`
3. Now repeat the above steps using this `src/dosbox` executable.
4. Notice that the player honors the up-key to jump obstacles at the right time.

#### Known affected games:

'7CitGold/Seven Cities of Gold, The (1985).bat'
'ActionFi/Action Fighter (1989).bat'
'Bang1994/Bang (1994).bat'
'BCsQuest/BC'\''s Quest for Tires (1984).bat'
'BioMenac/Bio Menace (1993).bat'
'BobMC1/Bob Morane - Chevalerie 1 (1987).bat'
'BoJackso/Bo Jackson Baseball (1991).bat'
'Brkoff89/Breakoff (1989).bat'
'BruceLee/Bruce Lee (1987).bat'
'BStorm/Battlestorm (1990).bat'
'calacub2/Calamari Cubes 2 (1998).bat'
'cospace/Cosmic Spacehead (1993).bat'
'DemonAtt/Demon Attack (1983).bat'
'Fluf/Fluf! (1989).bat'
'FreddyHa/Freddy Hardest (1988).bat'
'FredHard/Freddy Hardest in South Manhattan (1989).bat'
'Ghostbus/Ghostbusters (1986).bat'
'GhostsNG/Ghosts '\''N Goblins (1987).bat'
'Goal!/Goal! (1993).bat'
'Grem290/Gremlins 2 - The New Batch (1990).bat'
'Hball/HardBall! (1987).bat'
'HedHog84/Hedgehog (1984).bat'
'HelterSk/Helter Skelter (1989).bat'
'Inv1985/Invaders (1985).bat'
'Janosik/Janosik (1994).bat'
'Juno1st/Juno First (1984).bat'
'KickOff3/Kick Off 3 - European Challenge (1994).bat'
'KOff94/Kick Off 3 (1994).bat'
'LaEspada/Espada Sagrada, La (1990).bat'
'LaserCy/Laser Cycle (1983).bat'
'LastWarr/Last Warrior, The (1985).bat'
'LostTomb/Lost Tomb (1982).bat'
'MegaMan3/Mega Man 3 - The Robots are Revolting (1992).bat'
'Microsur/Microsurgeon (1984).bat'
'MightyBo/Mighty Bombjack (1990).bat'
'Miner204/Miner 2049er (1983).bat'
'MineShaf/Mine Shaft (1983).bat'
'MirageT/Mirage Thunder (1993).bat'
'MonsterB/Monster Bash (1993).bat'
'NetWars/NetWars (1993).bat'
'Ninja/Ninja (1986).bat'
'PacMan82/Pac Man (1982).bat'
'PitFigh/Pit-Fighter (1991).bat'
'PlayerMa/Player Manager 2 (1995).bat'
'Qbert/Qbert (1984).bat'
'Quaser90/Quasar (1990).bat'
'RickDang/Rick Dangerous (1989).bat'
'RiverRai/River Raid (1984).bat'
'SDEp1/Star Destroyer - Episode I (1986).bat'
'Sidewalk/Sidewalk (1987).bat'
'Snakes88/Snakes (1988).bat'
'Soldiers/Soldiers of Fortune (1994).bat'
'Solo/Solo (1989).bat'
'SpMinr83/Space Miner (1983).bat'
'Spyder/Spyder (1983).bat'
'StarRang/Star Rangers (1995).bat'
'Storm87/Storm (1987).bat'
'StreetFB/Street Football (1988).bat'
'StrShpAt/Star Ship Attack (1982).bat'
'Stryx/Stryx (1990).bat'
'SWTieCD/Star Wars TIE Fighter (Collector'\''s CD-ROM) (1995).bat'
'SWXWCD/Star Wars X-Wing (Collector'\''s CD-ROM) (1994).bat'
'ToadRoad/Toad Road (1990).bat'
'TotalCar/Total Carnage (1994).bat'
'TRAZ/TRAZ (1989).bat'
'Trojan/Trojan (1986).bat'
'Uridium/Uridium (1988).bat'
'Viper98/Viper (1998).bat'
'Voyager/Voyager I - Sabotage of the Robot Ship (1982).bat'
'VRSling/VR Slingshot (1994).bat'
'WingThun/Wings of Thunder (1994).bat'
'WldGolf2/World Golf 2 (1990).bat'
'Zapem/Zap'\''em (1982).bat'
'Zaxxon/Zaxxon (1984).bat'

### Download URL of affected game or software

https://github.com/dosbox-staging/dosbox-staging/files/11468755/bcquest.zip

### Your configuration

```ini
Embedded in zip.
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by kcgen** [comment]
@Nopey -- thank you for taking a look at this! 

Hopefully the above reproduction steps show the before and after effects. 

I did try to revert the original PR, however unfortunately it's been a couple years since then and we've slightly built on the mapper, so I wasn't able to easily  figure it out. Very much appreciate you taking a crack at it 👍

**Comment by kcgen** [comment]
@Nopey , just curious if you've managed any progress w/ this? 
(all good if not - just wanted to check in :sweat_smile: )

**Comment by kcgen** [comment]
> I've been able to reproduce the issue, and have been poking around in the mapper.
> Looking at whether important calls such as JOYSTICK_Move_X and JOYSTICK_Move_Y were happening; when broken, these functions never get called.

:+1: 

> Some (presumably desirable?) behaviors 7dbf5ce^ exhibits that main doesn't:
> 1. calls MAPPER_UpdateJoysticks every 20 ticks, even when no physical joysticks are present.
> 2. is willing to have joytype be JOY_2AXIS even without a physical stick (main's QueryJoysticks loves to clear joytype)
> 3. shows a non-existant physical joystick binding in the mapper, labelled as [missing joystick]

Nicely described. yes, all three (now missing) behaviors are the basis of the regression.
- Normally users without a physical joystick could still manually set `joysticktype = <something>` (1) and be presented with that joystick's enabled buttons in the mapper UI (2) in which to assign keystrokes and pump those events over to the emulated DOS joystick (3)

> I've reinstanted behaviors 1 and 2 in [my branch](https://github.com/Nopey/dosbox-staging/tree/fix-emulated-joysticks),

Great progress, @Nopey !

> I think 3's absence points towards the joystick still not being fully created/real without a physical controller

Yes, definitely. It seems like some initialization is now missing (in main) or perhaps that initialization is being cleared or reset.

I did a bisect-revert (repeat) process working forward to try to isolate the impacting commits on [`kc/reverts-for-nopey-1`](https://github.com/dosbox-staging/dosbox-staging/compare/kc/reverts-for-nopey-1), one revert that might be interesting is https://github.com/dosbox-staging/dosbox-staging/commit/dd2df39ff5354ede31f70e0c87f7eb343a82f31e, which removed the two calls to `ClearAllBinds();`.  Please ignore this otherwise.. it's a bit of a dog's breakfast!

> Am I missing something, or is this delete in CreateBindGroups extraneous

If `mapper.sticks.stick[mapper.sticks.num_groups]` already holds a previously `new`d `CStickBindGroup`, then it's useful from a leak-prevention perspective.  But I'm not seeing the `delete` pattern applied to the other joytype cases, so this could be a critical regression too - worth knocking it out and seeing if that prior object bind group should infact live on (in parallel).

**Comment by Nopey** [comment]
Adding this debug print to the top of `CContinuousEvent::ActivateEvent` illustrates the sticky keypress issue:
```c
		if(ev_trigger && activity > 0 && strncmp(GetName(), "jaxis",5) == 0)
			// this should only happen if two keys are bound to the same axis.
			printf("### ### incrementing %s's activity past %d\n", GetName(), (int)activity);
```

In both `main` and 7dbf5ce^, this shows that the key repeat events are erroneously incrementing the `activity` field; this causes the emulated joystick axis to continue to be held after the key is released.
As this isn't a regression caused by 7dbf, I think it should be a followup issue; I'm opening an MR that returns the 3 behaviors I described above
